File: Talleres/Taller 2/BinarioInversoDyV.py
from math import pow
#Convert an integer number into binary list
def DecimalToBinary(numero_decimal):
    binario = []
    while numero_decimal != 0:
        binario.insert(0, str(numero_decimal % 2))
        numero_decimal //= 2
    return (binario)

#Invert a chain given a list begin and end
def InvertChain(S,l,r):
    if l==r:
        return S[l]

    q = (l+r)//2
    left = InvertChain(S,l,q)
    right= InvertChain(S,q+1,r)
    result=[]

    #SWAP
    for i in range(len(right)):
        result.append(right[i])
    for i in range(len(left)):
        result.append(left[i])
    return result

#Convert a list of binary into an integer
def BinaryToDecimal(numero_binario):
    decimal = 0
    i=0
    inverted = InvertChain(numero_binario,0,len(numero_binario)-1)
    print("Inverted on binary", inverted)
    for i in range(len(inverted)):
        p = int(pow(int(inverted[i])*2,i))
        decimal = decimal + p
    return decimal

n = 345
print(n)
binario= DecimalToBinary(n)
print(binario)
binarioInverso = InvertChain(binario,0,len(binario)-1)
print(binarioInverso)
decimal= BinaryToDecimal(binarioInverso)
print(decimal)

File: Talleres/Taller 2/BinarioInversoDyV.cpp
#include <iostream>
#include <cmath>
using namespace std;

//Convert an integer number into binary list
string DecimalToBinary(int n)
{ 
    string binary = "";
    int i = 0;
    while (n > 0)
    {
        char c = n % 2 + '0';
        binary += c;
        n = n / 2;
        i++;
    }
    return binary;
}

//Invert a chain given a list begin and end
string InvertChain(string S, int l, int r)
{
    if(l==r)
    {
        string aux = "";
        aux+=S[l];
        return aux;
    }

    int q = (l+r)/2;
    string left = InvertChain(S,l,q);
    string right = InvertChain(S,q+1,r);
    string result = "";

    for(int i=0;i<right.size();i++)
    {
        result+=right[i];
    }
    for(int i=0;i<left.size();i++)
    {
        result+=left[i];
    }
    return result;
}

int BinaryToDecimal(string S)
{
    int decimal = 0;
    int i=0;
    string aux = InvertChain(S,0,S.size()-1);
    for(int i=0;i<aux.size();i++)
    {
        int c = aux[i]-'0';
        int p = c * pow(2,aux.size()-i-1);
        decimal+=p;
    }
    return decimal;
}

int main()
{
    int n=345;
    cout<<"DIVIDE AND CONQUER"<<endl;
    cout<<"Decimal: "<<n<<endl;
    string binary = DecimalToBinary(n);
    cout << "Binary: " << binary << endl;
    string inverted = InvertChain(binary,0,binary.size()-1);
    cout << "Inverted: " << inverted << endl;
    int decimal = BinaryToDecimal(inverted);
    cout << "Decimal: " << decimal << endl;
    return 0;
}